### PR TITLE
Update Stripe package

### PR DESCRIPTION
**Pull request by CodeAnimal**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://stripe.com/docs/api#card_object-customer
https://stripe.com/docs/api#pay_invoice
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

I've fixed a couple of small bugs in the definitions that were leading to errors for valid code.

### CODE REVIEW

**Comment by LinusU** [comment]
Not 100% about changing `IMetadata` to `any`, but the rest looks very good 👍 

The problem with `any` is that it allows you to pass in numbers, strings, etc. I'm not sure about the best way forward, but I think that I would recommend setting up your own types for the metadata, and then casting to those.

e.g.

```typescript
interface MyMetadata {
  foo: number
  bar: string
}

customer.cards.retrieve("card_17xMvXBoqMA9o2xkq6W5gamx").then((card) => {
  const metadata = card.customer.metadata as MyMetadata

  if (metadata.foo === 1337) {
    // ...
  }
})
```

**Comment by LinusU** [comment]
Haha, I forgot that I'm not allowed to make thumbs up 🙈 

Please hold off on merging this until we have discussed the `IMetadata` change...

File: types/stripe/index.d.ts
Comment (andy-ms) [inline types/stripe/index.d.ts]: Why is this necessary?

**Review by ghost — CHANGES_REQUESTED** [review]

File: types/stripe/index.d.ts
Comment (CodeAnimal) [reply types/stripe/index.d.ts]: Because of my proposed test case: [Line 248](https://github.com/CodeAnimal/DefinitelyTyped/blob/3dcc65aa22e52475e64cbe97742186f68615f83c/types/stripe/stripe-tests.ts#L248)

I agree, that `any` may not be the best type, but working with the metadata field can feel awkward and clumsy with the interface model, as the test example displays. It seems like having to jump through hoops, needing to cast the metadata field every time I want to read data from it. I'm not sure what the best solution is for this, what do we think?

Comment (CodeAnimal) [reply types/stripe/index.d.ts]: @LinusU @andy-ms Maybe this would work better:

```typescript
interface IMetadata {
    [x: string]: string | number;
}
```

This would cover all of these test cases:
```typescript
customer.metadata["test"] = "test";
customer.metadata["test"] = 1;
customer.metadata["test"] == "test";
customer.metadata["test"] == 1;
customer.metadata.test = "test";
customer.metadata.test = 1;
customer.metadata = {
    test: "test"
}
customer.metadata = null;
```

But wouldn't allow:
```typescript
customer.metadata = true;
customer.metadata = "test";
customer.metadata = 2;
customer.metadata.test = { nested: "Pojo"};
customer.metadata = {
    test: { nested: "Pojo" }
}
```

You would need to set the `TypeScript Version` header to 2.2 or above though. What do we think?

Comment (LinusU) [reply types/stripe/index.d.ts]: That seems very good!

Are you sure that you are allowed to send numbers to Stripe though? Or should it just be `{ [key: string]: string }`?

Comment (CodeAnimal) [reply types/stripe/index.d.ts]: I think the stripe-node library just stringifies the value to a url string, using the __qs__ package (see [here](https://github.com/stripe/stripe-node/blob/ab9fbfb72511afabfdb6b25ebe0ba3d8fef626e3/lib/StripeResource.js#L247) and [here](https://github.com/stripe/stripe-node/blob/ab9fbfb72511afabfdb6b25ebe0ba3d8fef626e3/lib/utils.js#L29)).

For example:
```typescript
metadata: {
    test: "123"
}
```
and
```typescript
metadata: {
    test: 123
}
```
should both stringify down to the url value: `"metadata[test]=123"` (at least the [docs](https://www.npmjs.com/package/qs#stringifying) seem to indicate that). You can see that in the [Stripe Metadata docs](https://stripe.com/docs/api/curl#metadata), if you compare the code examples for Node and cURL. Either way, I'm 99.999% confident that qs.stringify would treat the `123` and `"123"` as the same with regards to the output.

With that being the case, it seems valid to allow both a `number` or a `string`, as both are equally as valid.

Comment (CodeAnimal) [reply types/stripe/index.d.ts]: You might have a point on the __return value__ though. It will always return a string, maybe that should be taken into consideration.

**Comment by CodeAnimal** [comment]
I've committed the proposed changes for review.

**Comment by CodeAnimal** [comment]
@LinusU @andy-ms We could have two IMetadata interfaces, one for the return value (that would be the resources' `metadata` types):
```typescript
interface IMetadataReturn {
    [x: string]: string;
}
```
And one for setting/updating the metadata (e.g. when calling a `create` or `update` method):
```typescript
interface IMetadata {
    [x: string]: string | number;
}
```

**Comment by andy-ms** [comment]
If that's accurate, it sounds good. Waiting for review from @LinusU.

**Comment by CodeAnimal** [comment]
Reviewers can check the documentation for the changes in [my comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/23083#discussion_r163217898)

**Review by LinusU — APPROVED** [review]
Nice 👍

**Pull request merged**
